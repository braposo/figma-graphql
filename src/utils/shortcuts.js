const camelCase = require("lodash/camelCase");
const { nodeTypes } = require("../types/node");
const { getNodes } = require("./nodes");

const shortcutTypes = nodeTypes.map(type => {
    const convertedType = type === "CANVAS" ? "PAGE" : type;
    return camelCase(convertedType);
});

const mapTypeToFigmaType = nodeTypes.reduce(
    (acc, type) => {
        const figmaType = type === "CANVAS" ? "PAGE" : type;
        const key = `${camelCase(figmaType)}s`;
        return {
            ...acc,
            [key]: type,
        };
    },
    {
        children: "ALL",
    }
);

const mapTypeToQuery = nodeTypes.reduce(
    (acc, type) => {
        const convertedType = type === "CANVAS" ? "PAGE" : type;
        const formattedType = camelCase(convertedType);
        const key = `${formattedType}s`;
        const returnType = formattedType.charAt(0).toUpperCase() + formattedType.slice(1);
        return {
            ...acc,
            [key]: `${key}(name: String): [${returnType}!]`,
        };
    },
    {
        children: "children(type: [NodeType], name: String): [Node!]",
    }
);

exports.generateResolversForShortcuts = () =>
    [...shortcutTypes, "children"].reduce((acc, resolverType) => {
        const key = resolverType === "children" ? resolverType : `${resolverType}s`;
        return {
            ...acc,
            [key]: (root, { type, name }) =>
                getNodes(root, mapTypeToFigmaType[key], { type, name }),
        };
    }, {});

exports.generateQueriesForShortcuts = () =>
    [...shortcutTypes, "children"]
        .map(type => {
            const key = type === "children" ? type : `${type}s`;
            return mapTypeToQuery[key];
        })
        .join("\n");
