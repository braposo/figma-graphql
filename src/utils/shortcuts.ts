import camelCase from "lodash/camelCase";
import { nodeTypes } from "../types/node";
import { getNodes } from "./nodes";

const shortcutTypes = nodeTypes.map(type => {
    const convertedType = type === "CANVAS" ? "PAGE" : type;
    return camelCase(convertedType);
});

const mapTypeToFigmaType = nodeTypes.reduce(
    (acc, type) => {
        const figmaType = type === "CANVAS" ? "PAGE" : type;
        const key = `${camelCase(figmaType)}s`;
        return {
            ...acc,
            [key]: type,
        };
    },
    {
        children: "ALL",
    }
);

const mapTypeToQuery = nodeTypes.reduce(
    (acc, type) => {
        const convertedType = type === "CANVAS" ? "PAGE" : type;
        const formattedType = camelCase(convertedType);
        const key = `${formattedType}s`;
        const returnType = formattedType.charAt(0).toUpperCase() + formattedType.slice(1);
        const nodeType = type === "STYLE" ? "StyleType" : "NodeType";

        return {
            ...acc,
            [key]: `${key}(type: [${nodeType}], name: String): [${returnType}!]`,
        };
    },
    {
        children: "children(type: [NodeType], name: String): [Node!]",
    }
);

export const generateResolversForShortcuts = () =>
    [...shortcutTypes, "children"].reduce((acc, resolverType) => {
        const key = resolverType === "children" ? resolverType : `${resolverType}s`;
        return {
            ...acc,
            [key]: (root, { type, name }) =>
                getNodes(root, mapTypeToFigmaType[key], { type, name }),
        };
    }, {});

export const generateQueriesForShortcuts = () =>
    [...shortcutTypes, "children"]
        .map(type => {
            const key = type === "children" ? type : `${type}s`;
            return mapTypeToQuery[key];
        })
        .join("\n");
