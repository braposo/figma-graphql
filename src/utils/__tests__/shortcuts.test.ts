/* eslint-env jest */
import { generateResolversForShortcuts, generateQueriesForShortcuts } from "../shortcuts";

describe("Shortcuts", () => {
    test("can generateResolversForShortcuts", async () => {
        const resolvers = generateResolversForShortcuts();

        expect(resolvers).toMatchSnapshot();
    });

    test("can generateQueriesForShortcuts", async () => {
        const queries = generateQueriesForShortcuts();

        expect(queries).toMatchSnapshot();
    });
});
