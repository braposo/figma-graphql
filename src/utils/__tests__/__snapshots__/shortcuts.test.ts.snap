// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shortcuts can generateQueriesForShortcuts 1`] = `
"pages(type: [NodeType], name: String): [Page!]
frames(type: [NodeType], name: String): [Frame!]
groups(type: [NodeType], name: String): [Group!]
vectors(type: [NodeType], name: String): [Vector!]
booleanOperations(type: [NodeType], name: String): [BooleanOperation!]
stars(type: [NodeType], name: String): [Star!]
lines(type: [NodeType], name: String): [Line!]
ellipses(type: [NodeType], name: String): [Ellipse!]
regularPolygons(type: [NodeType], name: String): [RegularPolygon!]
rectangles(type: [NodeType], name: String): [Rectangle!]
texts(type: [NodeType], name: String): [Text!]
slices(type: [NodeType], name: String): [Slice!]
components(type: [NodeType], name: String): [Component!]
instances(type: [NodeType], name: String): [Instance!]
styles(type: [StyleType], name: String): [Style!]
children(type: [NodeType], name: String): [Node!]"
`;

exports[`Shortcuts can generateResolversForShortcuts 1`] = `
Object {
  "booleanOperations": [Function],
  "children": [Function],
  "components": [Function],
  "ellipses": [Function],
  "frames": [Function],
  "groups": [Function],
  "instances": [Function],
  "lines": [Function],
  "pages": [Function],
  "rectangles": [Function],
  "regularPolygons": [Function],
  "slices": [Function],
  "stars": [Function],
  "styles": [Function],
  "texts": [Function],
  "vectors": [Function],
}
`;
